### DISCUSSION TITLE
UserDefinedType stops working when upgrading from psycopg2 to psycopg3

### DISCUSSION THREAD

**Discussion by chekhovana** [sqlalchemy/sqlalchemy#11380] [Usage Questions]
The following code:
```
import sqlalchemy
from sqlalchemy import create_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from sqlalchemy.types import UserDefinedType
from dataclasses import dataclass
from typing import Tuple, Type, Any
import os


@dataclass(eq=True, frozen=True, slots=True)
class Location:
    lat: float
    lon: float


class LocationType(UserDefinedType):
    cache_ok = True

    @property
    def python_type(self) -> Type[Any]:
        return Location

    def get_col_spec(self):
        return "POINT"

    def bind_expression(self, bindvalue):
        return sqlalchemy.func.POINT(bindvalue, type_=self)

    def bind_processor(self, dialect):
        def process(value: Location | Tuple[float, float] | None) -> str | None:
            if value is None:
                return None
            if isinstance(value, tuple):
                value = Location(*value)
            return f"({value.lat},{value.lon})"
        return process

    def result_processor(self, dialect, coltype):
        def process(value: str) -> Location | None:
            if value is None:
                return None
            lat, lon = value.strip("()").split(",")
            return Location(float(lat), float(lon))
        return process


class Base(DeclarativeBase):
    pass


class TestPoint(Base):
    __tablename__ = 'test_point'
    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str]
    location: Mapped[Location] = mapped_column(LocationType)


for dialect in ('postgresql', 'postgresql+psycopg'):
    print('dialect:', dialect)
    try:
        db_url = f'{dialect}://{os.environ["DBURL"]}'
        engine = create_engine(db_url, echo=True)

        Base.metadata.create_all(engine)
        from sqlalchemy.orm import Session

        with Session(engine) as session:
            session.add(TestPoint(name='somewhere', location=Location(3, 4)))
            session.commit()
        print('success')
    except Exception as e:
        print('error: ', e)
```

works with psycopg 2.9, producing the output:
```
dialect: postgresql
...
2024-05-12 09:39:11,573 INFO sqlalchemy.engine.Engine INSERT INTO test_point (name, location) VALUES (%(name)s, POINT(%(location)s)) RETURNING test_point.id
2024-05-12 09:39:11,573 INFO sqlalchemy.engine.Engine [generated in 0.00012s] {'name': 'somewhere', 'location': '(3,4)'}
2024-05-12 09:39:11,577 INFO sqlalchemy.engine.Engine COMMIT
success
```
but generates error when switching to psycopg 3.1:

```
dialect: postgresql+psycopg
...
2024-05-12 09:39:11,697 INFO sqlalchemy.engine.Engine INSERT INTO test_point (name, location) VALUES (%(name)s::VARCHAR, POINT(%(location)s)) RETURNING test_point.id
2024-05-12 09:39:11,697 INFO sqlalchemy.engine.Engine [generated in 0.00011s] {'name': 'somewhere', 'location': '(3,4)'}
2024-05-12 09:39:11,699 INFO sqlalchemy.engine.Engine ROLLBACK
error:  (psycopg.errors.AmbiguousFunction) function point(unknown) is not unique
LINE 1: ... test_point (name, location) VALUES ($1::VARCHAR, POINT($2))...
                                                             ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
[SQL: INSERT INTO test_point (name, location) VALUES (%(name)s::VARCHAR, POINT(%(location)s)) RETURNING test_point.id]
[parameters: {'name': 'somewhere', 'location': '(3,4)'}]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

Comparing generated sql statements, I can see the followin difference - version 2 doesn't have field type declarations:

`INSERT INTO test_point (name, location) VALUES (%(name)s, POINT(%(location)s))`

whereas version 3 has explicit type for `name` field, but no type for `location` field:

`INSERT INTO test_point (name, location) VALUES (%(name)s::VARCHAR, POINT(%(location)s))`

I suggest that sql type for `location` field must also be declared, but I can't find the way to do it. In debug mode I see that method `LocationType::get_col_spec` isn't invoked.

### Environment
psycopg 3.1.19
psycopg-binary 3.1.19
psycopg2-binary 2.9.6
SQLAlchemy 2.0.30

Any help would be appreciated.

**Comment by chekhovana** [comment]
Setting `LocationType.render_bind_cast = True` solved the issue.

**Reply by CaselIT** [reply]
That's my suggestion too. psycopg3 uses server side binding and postgresql is very picky about types

**Reply by chekhovana** [reply]
Hi @CaselIT, thanks for the answer.

**Reply by chekhovana** [reply]
Hi @CaselIT,

I reopened the issue, because there is another problem with this user-defined type, during alembic migration:

`SAWarning: Did not recognize type 'point' of column 'location'`

Could you please suggest the way to fix it? 

I've made some debugging and it seems that problem arises here:
[https://github.com/sqlalchemy/sqlalchemy/blob/93cfb49572ac56bc320a09b82285bf8ef8cdff57/lib/sqlalchemy/dialects/postgresql/base.py#L3756](https://github.com/sqlalchemy/sqlalchemy/blob/93cfb49572ac56bc320a09b82285bf8ef8cdff57/lib/sqlalchemy/dialects/postgresql/base.py#L3756)

`ischema_names` doesn't have entry with `point` key, so `None` is returned.

**Reply by CaselIT** [reply]
I think that warning is also for psycog2. Can you confirm?

There aren't any extensions point at the moment to fix that issue. Try to add your class to ischema_names to see what happens.

If you just ignore it, does alembic break?

**Reply by chekhovana** [reply]
Hello @CaselIT 

> I think that warning is also for psycog2. Can you confirm?

Yes, you are right - the warning isn't dialect-specific, it's also reproducible with psycopg2.

> There aren't any extensions point at the moment to fix that issue. Try to add your class to ischema_names to see what happens.

Thanks for the advice, I'll try this

> If you just ignore it, does alembic break?

No, alembic doesn't break. Command `revision --autogenerate -m 'new'` generates the following output:

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
/app/configs/migrations/env.py:75: SAWarning: Did not recognize type 'point' of column 'location'
  context.run_migrations()
INFO  [alembic.autogenerate.compare] Couldn't determine database type for column 'defaults.location'
INFO  [alembic.autogenerate.compare] Couldn't determine database type for column 'scenario.location'
  Generating /app/configs/migrations/versions/b6a972aa7472_new.py ...  done
```

Migration script is generated successfully. So I think I'll try to fix it, but if I fail - it may be left as is, with warnings. I just thought I was doing something wrong, that's why I asked a question.

**Reply by CaselIT** [reply]
@zzzeek we periodicity have request to add custom types here. Maybe it makes sense to add a new dialect event that's called with the string returned by the reflection and the resolved type so that an user can customize it?

Shall I open an issue?

**Reply by zzzeek** [reply]
we have the column_reflect event.     But what we should do is add fields to ReflectedColumn that provides for the raw dialect-level information.    https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.DDLEvents.column_reflect

**Reply by CaselIT** [reply]
maybe, but that's not integrated in the type reflection logic, meaning that there is no way of suppressing the warning that's raised by the postgresql dialect.

it's also not dispatched when calling just the `get_columns` inspector method (not sure if it should there)

it's likely enough to just enhance the existing event that adding a completely new one though

**Reply by zzzeek** [reply]
OK dialect level event then

though this makes it hard if many different 3rd party libraries all have their own custom types they want to add reflection support

so maybe not.    putting values in ischema_names seems safest for now

**Reply by CaselIT** [reply]
I've added this one https://github.com/sqlalchemy/sqlalchemy/issues/11385.

Having a new event seems like not great, since it requires that every dialect implements it.
Maybe can document a "suppress_warning" keyword that could be used to suppress "I did not find the type" kind of warnings (at least on internal dialects)
